File: 40_dynamic_programming/40_3_restaurant_ratings.py
def restaurant_ratings(ratings: list)->int:
  dp = [0] * len(ratings)
  dp[0] = ratings[0]
  dp[1] = max(dp[0], ratings[1])

  for i in range(2, len(ratings)):
    dp[i] = max(dp[i-2] + ratings[i], dp[i-1])

  return dp[-1]

if __name__ == "__main__":
  # Example1
  ratings = [8, 1, 3, 9, 5, 2, 1]
  print(restaurant_ratings(ratings))
  # Example2
  ratings = [8, 1, 3, 7, 5, 2, 4]
  print(restaurant_ratings(ratings))

File: 40_dynamic_programming/40_3_restaurant_ratings.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int restaurant_ratings(std::vector<int>& ratings) {
  int n = ratings.size();
  std::vector<int> dp(n, 0);
  dp[0] = ratings[0];
  dp[1] = std::max(dp[0], ratings[1]);

  for (int i=2; i<n; i++) {
    dp[i] = std::max(dp[i-1], ratings[i]+dp[i-2]);
  }

  return dp[n-1];
}

int main() {
    // Example1
    std::vector<int> ratings1 = {8, 1, 3, 9, 5, 2, 1};
    std::cout << restaurant_ratings(ratings1) << std::endl;

    // Example2
    std::vector<int> ratings2 = {8, 1, 3, 7, 5, 2, 4};
    std::cout << restaurant_ratings(ratings2) << std::endl;

    return 0;
}

